frodo.js

```diff
@@ -1,25 +1,26 @@
 /**
```

Comment:
SCRIPT-The way you've refactored this complex method into simpler ones is great. It makes the code much more maintainable.